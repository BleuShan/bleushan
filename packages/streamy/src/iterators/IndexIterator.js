import {invariant} from '../utils/invariant.js'
import {isValidArrayLike} from '../utils/isValidArrayLike.js'

export class IndexIterator {
  #index = 0
  #source
  #onComplete

  constructor(source, onComplete) {
    invariant({
      condition: isValidArrayLike(source),
      message: 'expected source have a valid ArrayLike object',
      errorType: TypeError
    })
    this.#source = source
    this.#onComplete = onComplete
  }

  [Symbol.iterator]() {
    return this
  }

  next() {
    const result = {
      done: this.#index > this.#source.length - 1
    }

    if (result.done && this.#onComplete != null) {
      this.#onComplete()
    }

    if (!result.done) {
      result.value = this.#source[this.#index]
      this.#index += 1
    }

    return result
  }
}
