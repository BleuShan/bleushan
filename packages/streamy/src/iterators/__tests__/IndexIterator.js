import IndexIterator from '../IndexIterator.js'

describe('IndexIterator', () => {
  describe.each`
    name                          | source
    ${'a valid array'}            | ${[1, 2, 3]}
    ${'a valid arrayLike object'} | ${{0: 0, 1: 1, 2: 2, length: 3}}
    ${'a string'}                 | ${'test'}
  `('when called on $name', ({source}) => {
    const construct = () => new IndexIterator(source)

    it('should return an IndexIterator', () => {
      const result = construct()
      expect(result).toBeInstanceOf(IndexIterator)
    })

    it('should yield the same values as the source', () => {
      expect(Array.from(construct())).toEqual(Array.from(source))
    })
  })

  describe.each`
    name                          | source
    ${'a valid array'}            | ${[1, 2, 3]}
    ${'a valid arrayLike object'} | ${{0: 0, 1: 1, 2: 2, length: 3}}
    ${'a string'}                 | ${'test'}
  `('when called on $name with a completion callback', ({source}) => {
    const callback = jest.fn()
    const iterator = new IndexIterator(source, callback)

    it('should not have called the callback', () => {
      expect(callback).not.toHaveBeenCalled()
    })

    it('should not have called the callback while traversing', () => {
      // eslint-disable-next-line no-unused-vars
      for (const _ of iterator) {
        expect(callback).not.toHaveBeenCalled()
      }
    })

    it('should have call the callback', () => {
      expect(callback).toHaveBeenCalledTimes(1)
    })
  })

  describe.each`
    name                             | source
    ${'an invalid arrayLike object'} | ${{0: 0, 1: 1, 2: 2, length: 3.5}}
    ${'a number'}                    | ${1}
    ${'a NaN'}                       | ${NaN}
    ${'a undefined'}                 | ${undefined}
    ${'a null'}                      | ${null}
    ${'an object'}                   | ${{value: null}}
  `('when called on $name', ({source}) => {
    const construct = () => new IndexIterator(source)
    it('should throw the appropriate error', () => {
      expect(construct).toThrowErrorMatchingSnapshot()
    })
  })
})
