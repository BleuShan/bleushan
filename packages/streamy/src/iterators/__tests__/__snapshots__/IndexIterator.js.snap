// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexIterator when called on a NaN should throw the appropriate error 1`] = `"expected source to be a valid ArrayLike object"`;

exports[`IndexIterator when called on a null should throw the appropriate error 1`] = `"expected source to be a valid ArrayLike object"`;

exports[`IndexIterator when called on a number should throw the appropriate error 1`] = `"expected source to be a valid ArrayLike object"`;

exports[`IndexIterator when called on a undefined should throw the appropriate error 1`] = `"expected source to be a valid ArrayLike object"`;

exports[`IndexIterator when called on an invalid arrayLike object should throw the appropriate error 1`] = `"expected source to be a valid ArrayLike object"`;

exports[`IndexIterator when called on an object should throw the appropriate error 1`] = `"expected source to be a valid ArrayLike object"`;
